docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+    `e` is false.
+
+*   A logical or expression `e` like `e1 || e2` shows that `v` has type T if
+    either `e1` or `e2` shows that when they are false if `e` itself is false.
+    For example, in `o is String || itsMonday`, we know `o` must be a String if
+    the entire `||` expression evaluated to false.
+
+## Extending Scope
+
+### "Else" Scope
+
+The specification defines three scopes where promotion can apply:
+
+*   The "then" clause of a conditional expression (16.20):
+
+    ```dart
+    o is String ? oPromotedToStringHere : notHere
+    ```
+
+*   The right operand of a logical `&&` expression (16.22):
+
+    ```dart
+    o is String && oPromotedToStringHere
+    ```
+
+    This is because short-circuiting means the right operand is only evaluated
+    when the left is true.
+
+*   The "then" clause of an if statement (17.5):
+
+    ```dart
+    if (o is String) oPromotedToStringHere;
+    ```
+
+All of these correspond to code that is only reached if a condition expression
+evaluates to true. We extend this to cover facts that are known when an
+expression evaluates to false.
+
+*   In a conditional expression of the form `e1 ? e2 : e3`, we know that if `e1`
+    shows that `v` has type T when `e1` is false, then `v` is known to be of
+    type T in `e3`:
+
+    ```dart
+    o is! String ? notHere : oPromotedToStringHere
+    ```
+
+*   In a logical boolean expression of the form `e1 || e2`, if `e1` shows `v`
+    has type T when `e1` is false, then `v` is known to be of type T in `e2`:
+
+    ```dart
+    o is! String || oPromotedToStringHere
+    ```
+
+    This is because short-circuiting means the right operand is only evaluated
+    when the left is false.
+
+*   In an if statement, if the condition shows that `v` has type T when the
+    condition is false, then `v` is known to be of type T in the else clause.
+
+    ```dart
+    if (o is! String) notHere; else oPromotedToStringHere;
+    ```
+
+### Loop Scopes
+
+Much like if statements, we extend C-style for and while loops to promote inside
+their bodies when the condition is true and the condition expression shows a `v`
+has type T:
+
+```dart
+while (o is String) oPromotedToStringHere;
+
+for (; o is String;;) oPromotedToStringHere;
+```
+
+### Scopes After Exits
+
+The last common case where users expect type promotion to work but currently
+doesn't is in code that can only be reached if a type test succeeds because
+earlier code will cause the code to exit, as in:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) throw "Not a string.";
+  return o.length; // Should be promoted here.
+}
+```
+
+We say a statement "cannot complete normally" if the statement is one of:
+
+*   A return statement
+*   A rethrow statement
+*   An expression statement whose expression is a throw expression
+*   An if statement with an else where neither statement can complete normally
+*   A block whose last statement cannot complete normally
```

Comment:
I would like full control flow analysis too, *eventually*. But I'd like to not go there yet because that's going to be a *lot* of work. Can we stick with a smaller subset that covers the common cases in this proposal and then refine it later?

I think we will end up revisiting promotion anyway to handle non-nullable types and possible intersection and union types. That would be a good time to do the full flow analysis.